plugins/wacom-usb/fu-wac-plugin.c

```diff
@@ -33,6 +33,41 @@ fu_wac_plugin_write_firmware(FuPlugin *plugin,
 	return fu_device_write_firmware(device, blob_fw, progress, flags, error);
 }
 
+static gboolean
+fu_wac_plugin_composite_prepare(FuPlugin *self, GPtrArray *devices, GError **error)
+{
+	for (guint i = 0; i < devices->len; i++) {
+		FuDevice *device = g_ptr_array_index(devices, i);
+		if (FU_IS_WAC_DEVICE(device)) {
+			g_autoptr(FuDeviceLocker) locker = fu_device_locker_new(device, error);
+			if (locker == NULL)
+				return FALSE;
+			g_info("switching main device to flash loader");
+			if (!fu_wac_device_switch_to_flash_loader(FU_WAC_DEVICE(device), error))
+				return FALSE;
+		}
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_wac_plugin_composite_cleanup(FuPlugin *self, GPtrArray *devices, GError **error)
+{
+	for (guint i = 0; i < devices->len; i++) {
+		FuDevice *device = g_ptr_array_index(devices, i);
+		if (FU_IS_WAC_DEVICE(device)) {
+			g_autoptr(FuDeviceLocker) locker = fu_device_locker_new(device, error);
+			if (locker == NULL)
+				return FALSE;
+			g_info("resetting main device");
+			fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+			if (!fu_wac_device_update_reset(FU_WAC_DEVICE(device), error))
+				return FALSE;
```

Comment:
If this loop has to terminate early because of an error, won't it leave the remaining devices stuck in flash loader mode?

Reply:
Hmm, good question.  I think the answer is that it's not a problem -- as the devices in the loop will be FuWacDevice (for the main firmware) and FuWacModule (for all the other stuff) -- which means we only do *anything* to the main firmware and there's only one thing that can go wrong.

Reply:
Ah, okay. I was under the impression that this loop could potentially iterate over multiple physical devices (e.g. if you had both a Cintiq and an Intuos plugged in), which could be problematic. If the loop is instead always over the constituents of a single physical device, then there would indeed be no problem.

Reply:
The only way we'd do the Cintiq and Intuos in the same "transaction" is if you put the firmware for both in the same cab file, and if both needed updating. I think that's quite unlikely.